### PR TITLE
Various library work

### PR DESCRIPTION
**Pull request by soc**
(see individual commits)

### CODE REVIEW

File: dora/stdlib/Iterator.dora [hunk 1]
```diff
@@ -0,0 +1,4 @@
+trait Iterator {
+  fun hasNext() -> Bool;
+  fun next() -> Self;
```
Comment (soc) [inline dora/stdlib/Iterator.dora]: I think with this return type the impls for `IntRangeIter` and `CodepointIterator` should not compile, but they do.

I guess we are missing a check there, @dinfuehr?

Comment (dinfuehr) [reply dora/stdlib/Iterator.dora]: This is indeed surprising, most likely a bug.

Comment (soc) [reply dora/stdlib/Iterator.dora]: Shall I remove the commit from this PR for now?

Comment (dinfuehr) [reply dora/stdlib/Iterator.dora]: Yes.

Comment (soc) [reply dora/stdlib/Iterator.dora]: Ah, maybe better: I kept the file move and just undid the CodepointIterator impl and the result type change. Also acceptable?

Comment (soc) [reply dora/stdlib/Iterator.dora]: Repro for bug is https://github.com/dinfuehr/dora/issues/164

**Pull request merged**